backfiller/backfiller/main.py

```diff
@@ -129,159 +114,308 @@ def get_remote_segment(base_dir, node, stream, variant, hour, missing_segment,
 
 
-def backfill(base_dir, stream, variants, hours=None, nodes=None):
-	"""Loop over nodes backfilling from each.
+def list_hours(node, stream, variants, start=None):
+	"""Return a list of all available hours from a node.
 
-	Backfill from node/stream/variants to base_dir/stream/variants for each node
-	in nodes. If nodes is None, use get_nodes() to get a list of nodes to
-	backfill from. By default all hours are backfilled. If backfilling from a
-	node raises an exception, this just goes onto the next node."""
-
-	if nodes is None:
-		nodes = get_nodes()
-
-	#ideally do this in parallel
-	for node in nodes:
-		try:
-			backfill_node(base_dir, node, stream, variants, hours)
-		except Exception:
-			logging.exception("Error while backfilling node {}".format(node))
+	List all hours available from node/stream for each variant in variants
+	ordered from newest to oldest.
+	
+	Keyword arguments:
+	start -- If a datetime return hours after that datetime. If a number,
+	return hours more recent than that number of hours ago. If None (default),
+	all hours are returned."""
 
+	hour_lists = [list_remote_hours(node, stream, variant) for variant in variants]
+	hours = list(set().union(*hour_lists))
+	hours.sort(reverse=True) #latest hour first
 
-def is_iterable(x):
-	"""Test whether input is iterable."""
-	try:
- 		iter(x)
-	except TypeError:
-		return False
-	return True
 
+	if start is not None:
+		if not isinstance(start, datetime.datetime):
+			start = datetime.datetime.utcnow() - datetime.timedelta(hours=start)
+		hours = [hour for hour in hours if datetime.datetime.strptime(hour, HOUR_FMT) > start]
+	
+	return hours
+
+
+class BackfillerManager(object):
+	"""Manages BackfillerWorkers to backfill from a pool of nodes.
+
+	The manager regularly calls get_nodes to an up to date list of nodes. If no
+	worker exists for a node in this list or in the static_node list, the
+	manager starts one. If a worker corresponds to a node not in either list,
+	the manager stops it. If run_once, only backfill once."""
+
+	NODE_INTERVAL = 300 #seconds between updating list of nodes
+
+	def __init__(self, base_dir, stream, variants, static_nodes=[], start=None,
+			run_once=False, node_file=None, node_database=None, localhost=None):
+		"""Constructor for BackfillerManager.
+		Creates a manager for a given stream with specified variants. If
+		static_nodes is None, manager"""
+		self.base_dir = base_dir
+		self.stream = stream
+		self.variants = variants
+		self.static_nodes = static_nodes
+		self.start = start
+		self.run_once = run_once
+		self.node_file = node_file
+		self.node_database = node_database
+		self.localhost = localhost
+		self.stopping = gevent.event.Event()
+		self.logger = logging.getLogger("BackfillerManager({})".format(stream))
+		self.workers = {} # {node url: worker}
+
+	def stop(self):
+		"""Shut down all workers and stop backfilling."""
+		self.logger.info('Stopping')
+		for node in self.workers.keys():
+			self.stop_worker(node)
```

Comment:
I think we should leave it up to the `run()` function to stop the workers.
My concern is that if `stop()` is called while `run()` is halfway through, say, fetching nodes, then `run()` is now going to START those workers again (since they're no longer in `self.workers`), plus other similar issues. The best way to avoid race conditions on shared state like that is to only ever mutate that state from a canonical place, which in this case would be `run()`. So what this would look like would be that at the end of our `run` loop, we'd stop all workers after exiting the loop.

Reply:
Ok, after looking at the interaction between this, the `else: self.stop()` in `run()`, and the run_once behaviour, I think I understand.

The virtue of stopping the workers in `stop()` is that it will stop them even if we exited the `run()` loop due to `run_once`, and are currently waiting for them to finish.

The extra `stop()` in `run()` that is only called if we're already stopping is to shut down any extra errant workers caused by the race.

The race between this and the `run()` loop is probably best resolved by adding a check to `start_worker`, something like:
```
if self.stopping.is_set():
  logging.debug("Refusing to create new worker because we're stopping")
  return
```

Then we no longer need the extra `stop()` in run.


Some comments explaining this are probably also in order.

Reply:
Fixed this by adding the check to start and added some comments.